### PR TITLE
Ability to put a server in Down state to maintenance

### PR DESCRIPTION
**Pull request by ravening**
## Description

Adds a new global setting set.host.down.to.maintenance which when set
to true will allow setting a host in downstate into maintenance mode
so that when the host comes back up, it will be automatically set to
maintenance mode and the resource state will not be set to Enabled.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

1, Power off one hypervisor so that it enters DOWN state
2. set the global setting `set.host.down.to.maintenance` to true
3. Enable maintenance mode on that host
4. Start the hypervisor
5. When it comes back up, it will be in maintenance mode

### CODE REVIEW

**Review by weizhouapache — APPROVED** [review]

**Comment by GabrielBrascher** [comment]
@ravening what do you think PR https://github.com/apache/cloudstack/pull/4111? Does that PR fit the context of this PR?

**Comment by ravening** [comment]
> @ravening what do you think PR https://github.com/apache/cloudstack/pull/4111? Does that PR fit the context of this PR?

@GabrielBrascher looked at it and have few questions


The declare degrade API works only if host is in disconnected state? 
What if we set host to maintenance mode and after that for some reason it goes down, we need to execute the api.
But with global setting we don't need to execute another API.


Also to declare host as not dead, we need to call another API and then cancel maintenance mode but with global setting , you need to just cancel maintenance mode

Apart from these two extra steps, both are doing same functionality

**Comment by GabrielBrascher** [comment]
@ravening the declare host as degraded API command would serve to put a problematic host on a special state that would allow to mark the VMs as stopped and trigger HA to start VMs (when service offering has _haenabled=true_) on another hosts.

Your implementation is nice and covers some interesting ground. The API command requires intervention from ADMIN with care. This is because if the host is Disconnected but there are VMs running in it and writing on storage it may cause corruption if VMs start on another host and write on same storage. Therefore that command should only be executed if the host is down.

API description: _Declare host as 'Degraded'. Host must be on 'Disconnected' or 'Alert' state. The ADMIN must be sure that there are no VMs running on the respective host otherwise this command might corrupted VMs that were running on the 'Degraded' host_.

Threfore, it cannot be automated. I think that both implementations cover different gaps on some corner cases.

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java [hunk 1]
```diff
@@ -1302,6 +1305,17 @@ public Host maintain(final PrepareForMaintenanceCmd cmd) {
             throw new CloudRuntimeException("Host is already in state " + host.getResourceState() + ". Cannot recall for maintenance until resolved.");
         }
 
+        if (SET_HOST_DOWN_TO_MAINTENANCE.value() && (host.getStatus() == Status.Down)) {
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1308]: @ravening what do you think of extracting lines 1308 - 1318 to a method? With that, it would be possible to cover these cases with JUnit test methods?

**Review by GabrielBrascher — COMMENTED** [review]
Thanks for the PR @ravening. The code and feature are looking good.

I am not against merging it the way it is; however, a big plus would be adding a couple of test cases covering the added logics.

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java (hunk 1 above)
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1308]: @ravening consider the value in zone, as the config scope is specified as Zone.

```suggestion
        if (SET_HOST_DOWN_TO_MAINTENANCE.valueIn(host.getDataCenterId()) && (host.getStatus() == Status.Down)) {
```

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java [hunk 2]
```diff
@@ -1302,6 +1305,17 @@ public Host maintain(final PrepareForMaintenanceCmd cmd) {
...
+            if (host.getResourceState() == ResourceState.Enabled) {
+                _hostDao.updateResourceState(ResourceState.Enabled, ResourceState.Event.AdminAskMaintenance, ResourceState.PrepareForMaintenance, host);
+                _hostDao.updateResourceState(ResourceState.PrepareForMaintenance, ResourceState.Event.InternalEnterMaintenance, ResourceState.Maintenance, host);
+                return _hostDao.findById(hostId);
+            } else if (host.getResourceState() == ResourceState.ErrorInMaintenance) {
+                _hostDao.updateResourceState(ResourceState.ErrorInMaintenance, ResourceState.Event.InternalEnterMaintenance, ResourceState.Maintenance, host);
+                return _hostDao.findById(hostId);
+            }
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1316]: @ravening can also consider state "_ResourceState.ErrorInPrepareForMaintenance_" ?

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java (hunk 1 above)
Comment (ravening) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1308]: > @ravening consider the value in zone, as the config scope is specified as Zone.
> 
> 
> 
> ```suggestion
> 
>         if (SET_HOST_DOWN_TO_MAINTENANCE.valueIn(host.getDataCenterId()) && (host.getStatus() == Status.Down)) {
> 
> ```

@sureshanaparti i don't think we need that because the host can exist only in one zone at any time but I can change the setting to zone scope instead of global scope

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java (hunk 2 above)
Comment (ravening) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1316]: > @ravening can also consider state "_ResourceState.ErrorInPrepareForMaintenance_" ?

@sureshanaparti no. This is applicable only for hosts in maintenance state

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1316]: @ravening any specific reason to skip "_ResourceState.ErrorInPrepareForMaintenance_", and consider "_ResourceState.ErrorInMaintenance_", to put the host in maintenance? What if the host is back Up when the resource state is "ErrorInPrepareForMaintenance" ?

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java (hunk 1 above)
Comment (sureshanaparti) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1308]: @ravening Currently, the config scope is Zone, do you want to mark that as Global scope?

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java (hunk 2 above)
Comment (ravening) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1316]: > @ravening any specific reason to skip "_ResourceState.ErrorInPrepareForMaintenance_", and consider "_ResourceState.ErrorInMaintenance_", to put the host in maintenance? What if the host is back Up when the resource state is "ErrorInPrepareForMaintenance" ?

@sureshanaparti  this code doesn't take care of that... This is only for hosts already in maintenance mode

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1316]: ok @ravening

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java (hunk 1 above)
Comment (ravening) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1308]: @sureshanaparti I dont think I need to make changes here as config is at zone level..setting in one zone doesnt mess up setting in other zones

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1308]: @ravening May be some confusion here. My point is, use the config value specified at the scope defined in the config parameter, else update the config scope accordingly.

The config "SET_HOST_DOWN_TO_MAINTENANCE" defined above uses Zone Scope. So, In order to pick the Zone specific value of that config, you have to use "_valueIn(<zoneId>)_", not _value()_ (this picks the global config value). If your use case is to set the config at Cloud level, then change Scope to Global.

**Comment by svenvogel** [comment]
@ravening small question. down means not enabled?

**Comment by ravening** [comment]
> @ravening small question. down means not enabled?

@svenvogel need not be. The host is still enabled but is in down state

**Comment by DaanHoogland** [comment]
@ravening can you look at the conflict?

**Comment by ravening** [comment]
@GabrielBrascher so whats your opinion on this pr? shall I close it and take yours further?

**Comment by GabrielBrascher** [comment]
@ravening your PR covers a different scenario, very similar though. From what I understand, this PR avoids a problematic host to get back from down to Enabled.

I don't see any problem in merging this one, but here follows a small summary on what #4111 does:

In summary, it presents an API command to manually fence a problematic host. The command declares a host as degraded moving it from Alert or Disconnected to "Degraded" state. All VMs that were running and had HA enabled will be started on another host, VMs with HA disabled will triansit from "Running" state (obviously they are not running, but CloudStack does not know that) to Stopped.

In summary, this implementation is a simple way of avoiding a problematic host to get back and does not require admins to take action. The other PR requires admins to execute an API command when they know a Disconnected/Alert host is down and has no running VMs.

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java (hunk 1 above)
Comment (GabrielBrascher) [reply server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1308]: This observation makes sense. Can you please take a look at _org.apache.cloudstack.framework.config.ConfigKey.valueIn()_ @ravening?

**Review by GabrielBrascher — APPROVED** [review]
Overall code LGTM, left a small comment regarding the config key being used with _value()_ vs _valueIn()_.

@DaanHoogland @svenvogel @rhtyd @andrijapanicsb what do you think of this funcionality?

**Comment by GabrielBrascher** [comment]
@ravening can you please fix the conflicts?

**Comment by ravening** [comment]
> @ravening can you please fix the conflicts?

@GabrielBrascher have resolved the conflicts

**Comment by GabrielBrascher** [comment]
Thanks, @ravening!

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@GabrielBrascher functionality makes sense and is guarded by a setting. 2 lgtms and test-report, so merging
